### PR TITLE
Allow to specify secrets labels and annotations from keycloakclient

### LINKED ISSUE #413: Allow to specify secrets labels and annotations from keycloakclient
### Description

We have several applications in different namespaces needing SSO. Those can take the oidc client secret as input from a k8s secret in the same namespace as the application.

I first tried to expand the keycloak-operator to look for keycloakclients in other namespaces (https://github.com/keycloak/keycloak-operator/pull/398), this was rejected because of incomplete implementation.

My next try is to copy the created secret using kubed. But kubed only copy secrets having a label (for example `kubed.appscode.com/sync: "kubernetes.io/metadata.name=grafana"`, ref: https://appscode.com/products/kubed/v0.13.0-beta.0/guides/config-syncer/intra-cluster/).

I'm proposing to extend the CRD. For example:

```yaml
apiVersion: keycloak.org/v1alpha1
kind: KeycloakClient
metadata:
  namespace: keycloak
  name: grafana-oidc-client-secret
  labels:
    app: sso
  annotations:
    argocd.argoproj.io/sync-options: SkipDryRunOnMissingResource=true
spec:
  realmSelector:
    matchLabels:
      app: sso
  client:
    clientId: grafana
    clientAuthenticatorType: client-secret
    standardFlowEnabled: true
    redirectUris:
    - https://grafana.example.org/*
    webOrigins:
    - '+' # redirectUris
    defaultClientScopes:
    - email
    - profile
    # V V V V V V V V V V V here V V V V V V V V V V V
    secretTemplate:
      metadata:
        labels:
          kubed.appscode.com/sync: "kubernetes.io/metadata.name=grafana"
    # Λ Λ Λ Λ Λ Λ Λ Λ Λ Λ Λ here Λ Λ Λ Λ Λ Λ Λ Λ Λ Λ Λ
``` 

The code changes are around:

https://github.com/keycloak/keycloak-operator/blob/77d126113104d45e29057f5c9e30c54f1a482763/pkg/model/client_secret.go#L13-L26

See also my WIP Ansible playbooks: https://gitlab.com/kubitus-project/kubitus-installer/-/merge_requests/364

### PR DESCRIPTION
**Pull request by sathieu**
## Related GH Issue
Closes #413

## Verification Steps

I've added unit tests.

1. Create a keycloakclient resource with the following spec:

```yaml
apiVersion: keycloak.org/v1alpha1
kind: KeycloakClient
metadata:
  namespace: keycloak
  name: grafana-oidc-client-secret
  labels:
    app: sso
  annotations:
    argocd.argoproj.io/sync-options: SkipDryRunOnMissingResource=true
spec:
  realmSelector:
    matchLabels:
      app: sso
  client:
    clientId: grafana
    clientAuthenticatorType: client-secret
    standardFlowEnabled: true
    redirectUris:
    - https://grafana.example.org/*
    webOrigins:
    - '+' # redirectUris
    defaultClientScopes:
    - email
    - profile
    # V V V V V V V V V V V here V V V V V V V V V V V
    secretTemplate:
      metadata:
        labels:
          kubed.appscode.com/sync: "kubernetes.io/metadata.name=grafana"
    # Λ Λ Λ Λ Λ Λ Λ Λ Λ Λ Λ here Λ Λ Λ Λ Λ Λ Λ Λ Λ Λ Λ
``` 

2. The keycloak operator will create a secret like this:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: grafana-oidc-client-secret
  labels:
    kubed.appscode.com/sync: "kubernetes.io/metadata.name=grafana"
# ...
```

## Checklist:
- [x] Automated Tests
- [ ] [Documentation](https://github.com/keycloak/keycloak-documentation) changes if necessary
- [ ] (If required) Discussed on [Keycloak DEV](https://groups.google.com/forum/#!forum/keycloak-dev)

### CODE REVIEW

**Comment by sathieu** [comment]
@vmuzikar @pedroigor @stianst @abstractj Pleas review :pray:!

**Comment by abstractj** [comment]
@andreaTP @vmuzikar FYI

**Comment by andreaTP** [comment]
@sathieu the implementation seems to be correct regarding adding the `label`s/`annotation`s; I have no issues with the actual code.

I understand the purpose of this change but it looks a bit involved from a UX perspective, the workflow, as I understand it, looks like:
 - a user creates in one namespace the client-secret
 - the client-secret is propagated across namespaces
 - the deployments are re-started to pick up the new secret?
 
 Now I have some questions:
- what happens if you create multiple different secrets with this annotation? (e.g. with a GitOps pipeline)
- Keycloak deployments are restarted accordingly?
- [most important] can this approach be applied without the additional `kubed` operator dependency?

Thanks a lot in advance for the answer!

**Comment by andreaTP** [comment]
@sathieu fair answer in this case LGTM

**Review by andreaTP — APPROVED** [review]

**Comment by sathieu** [comment]
@andreaTP Thanks for approving this PR. Anything I can do to move it forward?

**Comment by andreaTP** [comment]
Hi @sathieu , I don't have rights to merge, we can ping @vmuzikar to review.

**Review by vmuzikar — CHANGES_REQUESTED** [review]
@sathieu Thank you for your contribution! The suggestion sounds reasonable to me, changes look good. Could you please add an E2E test for it? I'd like to have this covered not only by a unit test, just in case...

Also, the GH Actions didn't run for some reason. I'll try to close and re-open this PR.

**Comment by sathieu** [comment]
@vmuzikar I've added some e2e tests. Please unlock the workflows :pray:.

File: pkg/model/client_secret.go [hunk 1]
```diff
@@ -10,19 +10,27 @@ import (
...
 	}
+	if cr.Spec.Client.SecretTemplate != nil && cr.Spec.Client.SecretTemplate.Metadata != nil {
+		if cr.Spec.Client.SecretTemplate.Metadata.Labels != nil {
+			secret.Labels = cr.Spec.Client.SecretTemplate.Metadata.Labels
+		}
+		if cr.Spec.Client.SecretTemplate.Metadata.Annotations != nil {
+			secret.Annotations = cr.Spec.Client.SecretTemplate.Metadata.Annotations
+		}
+	}
+	secret.Labels["app"] = ApplicationName
```
Comment (vmuzikar) [inline pkg/model/client_secret.go:32]: Something similar also needs to be done here:
https://github.com/keycloak/keycloak-operator/blob/9c9e2b977ff475045ee50f3dd4dcfdc2985388f7/pkg/model/client_secret.go#L44

File: test/e2e/keycloak_client_test.go [hunk 2]
```diff
@@ -592,3 +596,40 @@ func keycloakClientScopeMappingsTest(t *testing.T, framework *test.Framework, ct
 
 	return nil
 }
+
+func keycloakClientSecretTemplateTest(t *testing.T, framework *test.Framework, ctx *test.Context, namespace string) error {
+	// create
+	client := getKeycloakClientCR(namespace, false)
+	client.Spec.Client.SecretTemplate = &keycloakv1alpha1.SecretTemplate{
+		Metadata: &keycloakv1alpha1.SecretTemplateMetadata{
+			Labels: map[string]string{
```
Comment (vmuzikar) [inline test/e2e/keycloak_client_test.go:605]: We should also test annotations can be set.

**Comment by tomhuang12** [comment]
Hoping to have this. @sathieu are you still working on this?

**Comment by sathieu** [comment]
@tomhuang12 I still need this, but while the unit tests are passing, the e2e tests are failing with "400 Bad request".

I won't work on this soon, feel free to take it.

**Comment by vmuzikar** [comment]
@sathieu Thank you for your response. I'm going to close this for now. Feel free to reopen or open a new PR when you are able to finish it.

**Pull request closed without merging**